components/renku_data_services/authz/authz.py

```diff
@@ -580,7 +580,7 @@ def _add_project(self, project: Project) -> _AuthzChange:
             object=(
                 _AuthzConverter.user_namespace(project.namespace.id)
                 if project.namespace.kind == NamespaceKind.user
-                else _AuthzConverter.group(project.namespace.underlying_resource_id)
+                else _AuthzConverter.group(cast(ULID, project.namespace.underlying_resource_id))
```

Comment:
Why do we need this cast?

Reply:
because `undelying_resource_id` is an ULID for groups but a str keycloak id for user namespaces. so we know it is ULID here due to the if/else on `kind`, but mypy does not know this.
somewhat related, we should discuss at some point on if we want to use our own user ids or if we want to keep using keycloak ones directly.